src/build.rs

```diff
@@ -263,3 +265,323 @@ fn build_nasm(root: &Path, vendor_dir: &Path, out_dir: &Path, target_arch: &str,
+                let end_idx1 = decl.find('(');
+                let end_idx = match (end_idx0, end_idx1) {
+                    (Some(v), None) => v,
+                    (None, Some(v)) => v,
+                    (Some(i), Some(j)) => i.min(j),
+                    (None, None) => panic!(),
+                };
+                decl["fn ".len()..end_idx].to_string()
+        };
+        match name.as_str() {
+            | "jpeg_float_add_quant_table"
+                | "jpeg_calc_jpeg_dimensions"
+                | "jsimd_fdct_ifast"
+                | "jsimd_can_rgb_ycc"
+                | "jsimd_can_fdct_ifas"
+                => continue,
+
+            _ => (),
+        }
+        let lifetime = decl.find('<').map(|v| decl[v+1..decl.find('>').unwrap()].to_string());
+        let args: Vec<(String,String)> = decl.find('(')
+                .map(|v| &decl[v+1..decl.find(')').unwrap()]).unwrap().split(',')
+            .filter(|s| !s.trim().is_empty())
+            .map(|s|{
+                let mut s = s.split(':');
+                let arg_name = s.next().unwrap();
+                let ty_name = s.next().unwrap();
+                (arg_name.trim().to_string(), ty_name.trim().to_string())
+            }).collect();
+        let ret_ty = decl.rfind("->").map(|s| decl[s+2..].trim().to_string());
+        decls.push( Decl { name, is_pub, doc, attrs, args, ret_ty, lifetime });
+    }
+
+    decls
+}
+
+fn generate_cxx_wrappers(decls: &Vec<Decl>) -> Vec<String> {
+    decls.iter().map(|d| generate_cxx_wrapper(d)).collect()
+}
+
+fn generate_rust_wrappers(decls: &Vec<Decl>) -> Vec<String> {
+    decls.iter().map(|d| generate_rust_wrapper(d)).collect()
+}
+
+fn generate_rust_wrapper(decl: &Decl) -> String {
+    format!(r#"
+    {doc}
+    {attrs}
+    {pub_vis} unsafe fn {wrapper_name}{lifetime}({params}) -> {ret_ty} {{
+        extern "C" {{
+            fn {c_wrapper_name}{lifetime}({params}) -> WrapperResult<{ret_ty}>;
+        }}
+        let result = {c_wrapper_name}({arg_ids});
+        if result.discriminant == 0 {{
+           result.value.value
+        }} else {{
+           let WrapperError{{ buf, len, cap }}: WrapperError = result.value.error;
+           panic!(String::from_raw_parts(buf, len, cap))
+        }}
+    }}
+    "#,
+            doc = decl.doc.join("\n"),
+            attrs = decl.attrs.join("\n"),
+            pub_vis = if decl.is_pub { "pub" } else { "" },
+            wrapper_name = decl.name,
+            lifetime = decl.lifetime.clone().map(|l| format!("<{}>", l)).unwrap_or_default(),
+            params = decl.args.iter().map(|(id, ty)| format!("{}: {}", id, ty)).collect::<Vec<_>>().join(","),
+            c_wrapper_name = format!("{}_cxx", decl.name),
+            arg_ids = decl.args.iter().map(|(id, _)| id.clone()).collect::<Vec<_>>().join(","),
+            ret_ty = decl.ret_ty.clone().unwrap_or("()".to_string()),
+    )
+}
+
+fn generate_cxx_wrapper(decl: &Decl) -> String {
```

Comment:
This is the other half of the work. It generates the C++ wrappers that get called from the Rust code, and which call the mozjpeg functions from C++. They use a try catch to catch all exceptions, but you could use a setjmp or errno based approach here without issues.